modules/node-shared/src/main/scala/org/tessellation/node/shared/domain/seedlist/HistoricalSnapshotOrdinals.scala

```diff
@@ -0,0 +1,121 @@
+package org.tessellation.node.shared.domain.seedlist
+
+import cats.data.{NonEmptyList, ValidatedNel}
+import cats.syntax.all._
+
+import scala.util.control.NoStackTrace
+
+import eu.timepit.refined.auto._
+import eu.timepit.refined.types.numeric.NonNegLong
+import fs2.data.csv.{CellDecoder, DecoderError}
+
+case class HistoricalSnapshotOrdinals(values: NonEmptyList[Range])
+
+case class InvalidOrdinal(ordinal: String) extends NoStackTrace {
+  override def getMessage = s"Invalid ordinal: $ordinal"
+}
+
+case class InvalidRangeFormat(range: List[String]) extends NoStackTrace {
+  override def getMessage = s"Invalid range format: ${range.mkString("-")}"
+}
+
+case class InvalidInput(str: String) extends NoStackTrace {
+  override def getMessage = s"Invalid input: $str"
+}
+
+case class ReversedRange(start: NonNegLong, end: NonNegLong) extends NoStackTrace {
+  override def getMessage =
+    s"Invalid range: '$start-$end'. The start value must not be greater than the end value."
+}
+
+object HistoricalSnapshotOrdinals {
+
+  implicit val optionalHistoricalSnapshotOrdinalsCellDecoder: CellDecoder[Option[HistoricalSnapshotOrdinals]] =
+    CellDecoder.stringDecoder.emap { s =>
+      parseInput(s)
+        .map(reduce)
+        .andThen(_.toNel match {
+          case Some(value) => value.validNel
+          case None        => InvalidInput(s).invalidNel
+        })
+        .map(HistoricalSnapshotOrdinals(_).some)
+        .leftMap { errs =>
+          val msg = errs.map(_.getMessage).toList.mkString("; ")
+
+          new DecoderError(msg = msg)
+        }
+        .toEither
+    }
+
+  private val valueDelimiter = "-"
+  private val entryDelimiter = ";"
+
+  type ValidationResult[A] = ValidatedNel[Throwable, A]
+
+  private def parseInput(input: String): ValidationResult[Seq[Range]] =
+    input
+      .split(entryDelimiter)
+      .toList
+      .map(_.split(valueDelimiter).map(_.trim).toList)
+      .traverse(parseRange)
+
+  private def parseRange(range: List[String]): ValidationResult[Range] =
+    range match {
+      case start :: Nil =>
+        parseOrdinal(start).map(Range(_, NonNegLong.MaxValue))
+      case start :: end :: Nil =>
+        (parseOrdinal(start), parseOrdinal(end)).mapN {
+          case (start, end) =>
+            Range(start = start, end = end)
+        }.andThen {
+          case r if r.valid => r.validNel
+          case r            => ReversedRange(r.start, r.end).invalidNel
+        }
+      case _ =>
+        InvalidRangeFormat(range).invalidNel[Range]
+    }
+
+  private def parseOrdinal(ordinal: String): ValidationResult[NonNegLong] =
+    Either
+      .catchOnly[NumberFormatException](ordinal.toLong)
+      .leftMap(_ => InvalidOrdinal(ordinal))
+      .flatMap { l =>
+        NonNegLong
+          .from(l)
+          .leftMap(_ => InvalidOrdinal(ordinal))
+      }
+      .toValidatedNel
```

Comment:
Alternatively you can avoid generating an exception by doing something like this
```
      Either
        .fromOption(ordinalStr.toLongOption, InvalidOrdinal(ordinalStr))
        .flatMap(a => NonNegLong.from(a).leftMap(_ => InvalidOrdinal(ordinalStr)))
        .toValidatedNel
```

Reply:
Changed to:

```scala
  private def parseOrdinal(ordinal: String): ValidationResult[NonNegLong] =
    Either
      .fromOption(ordinal.toLongOption, InvalidOrdinal(ordinal))
      .flatMap(a => NonNegLong.from(a).leftMap(_ => InvalidOrdinal(ordinal)))
      .toValidatedNel
```